Issue 1384: Huge performance issue between 6.5.2 and 6.11

bazhenov opened the pull request:
Fixes #1384: Huge performance issue between 6.5.2 and 6.11.

### Did you remember to?

- [x] Add test case(s)
- [x] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

krmahadevan commented on src/test/java/testhelper/PerformanceUtils.java:
```diff
@@ -0,0 +1,28 @@
+  private static ThreadMXBean THREAD_MX_BEAN;
+
+  static {
+    Object bean = ManagementFactory.getThreadMXBean();
```
Curious. Why get it first as an object, run an instance check and then perform the casting ? The signature of `getThreadMXBean()` states it would return a `ThreadMXBean`.

```java
public static ThreadMXBean getThreadMXBean() {
    return ManagementFactoryHelper.getThreadMXBean();
}
```

bazhenov replied:
`getThreadMXBean()` returns `java.lang.management.ThreadMXBean`, but we need `com.sun.management.ThreadMXBean` to be able to call `getThreadAllocatedBytes(long)` (it's a part of a private API). This impose potential problem. This code would work only on OpenJDK/Oracle JVM. I'm not sure if this is important.

krmahadevan replied:
Thanks for the clarification. I didn't notice the package names.. My bad!

krmahadevan commented on CHANGES.txt:
```diff
@@ -1,5 +1,6 @@
 Current
 Fixed: GITHUB-1493: Wrong exception msg when timeout on test (Krishnan Mahadevan)
+Fixed: GITHUB-1384: Huge performance issue between 6.5.2 and 6.11
```
Just a suggestion: You should add your name there, since you put in effort on analyzing this and fixing it :)

bazhenov replied:
You're right, thanks!

krmahadevan commented:
@bazhenov - As part of this PR we revived the primitive array comparisons for `byte[]` comparisons, but this would now require code changes for being benefitted from this method no ?

```java
```

bazhenov commented:
@krmahadevan 

> but this would now require code changes for being benefitted from this method no ?

As far as I understand no. Java linker will choose most specific version of `assertEquals()` method while loading class `TestLZFRoundTrip`. It'll be `assertEquals(byte[], byte[])` with new version of TestNG.

To be sure, I'll check if my understanding is right.

krmahadevan commented:
> As far as I understand no. Java linker will choose most specific version of assertEquals() method while loading class TestLZFRoundTrip. It'll be assertEquals(byte[], byte[]) with new version of TestNG.

True that. I confirmed it with this sample.

```java
public class SampleTestClass {
    public static void main(String[] args) {
        foo(new byte[] {});
        foo(new Byte[]{});
    }

    public static void foo(Object object) {
        System.out.println("object() variant called");

    }

    public static void foo(byte[] object) {
        System.out.println("primitive() variant called");

    }
}
```

```
primitive() variant called
object() variant called
```

bazhenov commented:
Ok, I need some more time to add other primitive-based methods to `Assert` as well as remove some duplication and dead code. Please post-pone merging this pull-request :)

bazhenov commented:
I've added `assertEquals()` for all 7 primitive types and moved away all reference and length checks in shared method to minimize code duplication. Shared method is reflection based, but all hot-path code (element comparison) is reflection-free.

krmahadevan commented on src/main/java/org/testng/Assert.java:
```diff
@@ -169,6 +169,271 @@ private static void assertArrayEquals(Object actual, Object expected, String mes
+   * @return {@code true} if {@code actualArray} and {@code expectedArray} are the same, {@code false} otherwise. If
+   * references are different and arrays length are different {@link AssertionError} is thrown.
+   */
+  private static boolean checkRefEqualityAndLength(Object actualArray, Object expectedArray, String message) {
```
Does this method need to return anything ? We aren't doing anything with the returned value. We could just have it be a `void` method instead.

bazhenov replied:
It returns boolean flag indicating reference equality of both `actual` and `expected` so we don't have to do element-wise comparison.

krmahadevan commented on src/main/java/org/testng/Assert.java:
```diff
@@ -169,6 +169,271 @@ private static void assertArrayEquals(Object actual, Object expected, String mes
+    }
+  }
+
+  static private void failOnArrayMismatch(Object actual, Object expected, int idx, String message) {
```
Instead of relying on one global method which again uses reflection behind the scenes, why not resort to something like this : 

```java
private static final String TEMPLATE = "Arrays differ firstly at element [%d]; Expected value is<%s>, Actual value was <%s>. ";

  static public void assertEquals(long[] actual, long[] expected, String message) {
    if (checkRefEqualityAndLength(actual, expected, message)) return;

    for (int i = 0; i < expected.length; i++) {
      if (expected[i] != actual[i]) {
        String msg = String.format(TEMPLATE, i, "%d", "%d") + message;
        fail(String.format(msg, expected[i], actual[i]));
      }
    }
  }
```

This would help us do-away with this method (Which uses the reflection api), and works directly with primitive types and has the re-usability as well.

bazhenov replied:
Nice idea. We indeed doesn't have to use reflection for reporting error. I think we doesn't have to do "double templating" if we just convert argument to strings in place. Something like this:

```java
private static final String TEMPLATE = "Arrays differ firstly at element [%d]; Expected value is<%s>, Actual value was <%s>. ";

  static public void assertEquals(long[] actual, long[] expected, String message) {
    if (checkRefEqualityAndLength(actual, expected, message)) return;

    for (int i = 0; i < expected.length; i++) {
      if (expected[i] != actual[i]) {
        fail(String.format(TEMPLATE, i, Long.toString(expected[i]), Long.toString(actual[i])));
      }
    }
  }
```

What do you think?

krmahadevan replied:
Yep.. That would be good.

juherr commented on src/main/java/org/testng/Assert.java:
```diff
@@ -169,6 +172,270 @@ private static void assertArrayEquals(Object actual, Object expected, String mes
+   * @param actual   the actual value
+   * @param expected the expected value
+   */
+  static public void assertEquals(byte[] actual, byte[] expected) {
```
`public static` otherwise sonar will complain.

bazhenov replied:
done

juherr commented on src/test/java/testhelper/PerformanceUtils.java:
```diff
@@ -0,0 +1,28 @@
+
+public class PerformanceUtils {
+
+  private static ThreadMXBean THREAD_MX_BEAN;
```
should be `final`

bazhenov replied:
Unfortunately, it's impossible. Field initialized in static block which disallows `final` modifier.

juherr commented:
Ping @cbeust

cbeust commented:
Can I get a high level overview of the problem and the fix? It's a bit hard to follow from the code alone.

Also, for future reference, I prefer to see cosmetic changes (such as swapping `public static`) in a separate PR for easier review.

krmahadevan commented:
@cbeust - TestNG was using the `Arrays.get()` for running equality assertions on primitive arrays which was adding to the execution time. The changeset basically reverts back this behavior in TestNG and causes TestNG to work with primitive arrays directly instead of using `Arrays.get()`.
